### PR TITLE
Add in option to fold EM Shower daughters in CAF-making

### PR DESCRIPTION
**Pull request by gputnam**

### CODE REVIEW

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+bool isShowerDaughter(const caf::SRTrueParticle &p) {
+  return isShowerDaughterProcess(p.start_process);
+}
+
+std::vector<caf::SRTrueParticle> caf::FoldEMShowerDaughters(const std::vector<caf::SRTrueParticle> &ps) {
+  std::vector<caf::SRTrueParticle> ret;
+
+  // Speed gain -- construct a map from track ID to particle for fast lookup
+  std::map<int, const caf::SRTrueParticle *> idmap;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:30]: Probably `std::unordered_map` is even faster? The default map and set classes are pretty terrible for speed and (especially) space efficiency.

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  // Do the rollup
+  for (const caf::SRTrueParticle &p: ps) {
+    if (!isShowerDaughter(p)) continue;
+
+    ret.push_back(p);
+    
+    // Not a Shower-Daughter!
+    // For these particles -- collect up the visible energy of the daughters
+    for (unsigned d: p.daughters) {
+      if (!idmap.count((int)d)) continue;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:44]: Make the IDMap use unsigned ints so you don't need this cast? Or does that just introduce one when you fill it? Sigh.

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  // Do the rollup
+  for (const caf::SRTrueParticle &p: ps) {
+    if (!isShowerDaughter(p)) continue;
+
+    ret.push_back(p);
+    
+    // Not a Shower-Daughter!
+    // For these particles -- collect up the visible energy of the daughters
+    for (unsigned d: p.daughters) {
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:43]: Spell out `unsigned int` in full

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      if (!idmap.count((int)d)) continue;
+      const caf::SRTrueParticle &d_p = *idmap.at((int)d);
+      if (!isShowerDaughter(d_p)) continue;
+
+      for (unsigned i_c = 0; i_c < 2; i_c++) {
+        for (unsigned i_p = 0; i_p < 3; i_p++) {
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:49]: This is my ignorance of the structure. `i_p` must be plane, what is `i_c`?

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  // Speed gain -- construct a map from track ID to particle for fast lookup
+  std::map<int, const caf::SRTrueParticle *> idmap;
+  for (const caf::SRTrueParticle &p: ps) {
+    idmap[p.G4ID] = &p;
+  }
+
+  // Do the rollup
+  for (const caf::SRTrueParticle &p: ps) {
+    if (!isShowerDaughter(p)) continue;
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:37]: Is this test backwards? We want to include all the non-daughters in the vector, and skip over all the daughters, right?

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx (hunk 3 above)
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:43]: Is this enough? Or do we potentially need to recurse through a whole tree of grand-daughters?

File: sbncode/CAFMaker/CAFMaker_module.cc [hunk 6]
```diff
@@ -800,6 +801,10 @@ void CAFMaker::produce(art::Event& evt) noexcept {
     }
   }
 
+  if (fParams.FoldEMShowerDaughters()) {
+    true_particles = caf::FoldEMShowerDaughters(true_particles);
+  }
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CAFMaker_module.cc:806]: Is this a no-op if the daughters are already folded at the geant level? That would actually be a good test one could do relatively simply using `diff_cafs`

File: sbncode/CAFMaker/CMakeLists.txt [hunk 7]
```diff
@@ -25,6 +26,7 @@ art_make_library( LIBRARY_NAME sbncode_CAFMaker
                   ${ROOT_X3d}
                   ${ROOT_BASIC_LIB_LIST}
                   caf_RecoUtils
+		  caf_FoldEMDaughters
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/CMakeLists.txt:29]: Indentation looks off?

File: sbncode/CAFMaker/FoldEMDaughters/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+art_make_library( LIBRARY_NAME caf_FoldEMDaughters
+                  SOURCE  FoldEMDaughters.cxx
+                  LIBRARIES 
+                    cetlib 
+		    sbnanaobj_StandardRecord
```
Comment (cjbackhouse) [inline sbncode/CAFMaker/FoldEMDaughters/CMakeLists.txt:5]: Indentation

**Review by cjbackhouse — COMMENTED** [review]
Various minor comments, but I'm also yet to be convinced that the overall algorithm is correct.

File: sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx (hunk 4 above)
Comment (cjbackhouse) [reply sbncode/CAFMaker/FoldEMDaughters/FoldEMDaughters.cxx:49]: Answering self. "cryostat", right? `iCryo` and `iPlane` would have helped me.

**Comment by jzennamo** [comment]
@gputnam is this intended for the upcoming (2022A) production release?

**Comment by gputnam** [comment]
This PR has been stale for a while -- I'm closing it.

**Pull request closed without merging**